File: y2018/python/day13.py
import enum


# i realize i could have just used raw characters... oh well...
class Track(enum.Enum):
    HORIZONTAL = enum.auto()
    VERTICAL = enum.auto()
    SLASH = enum.auto()
    BACKSLASH = enum.auto()
    PLUS = enum.auto()


class Cart(enum.Enum):
    RIGHT = 0
    DOWN = 1
    LEFT = 2
    UP = 3


CHAR_MAP = {
    '-': Track.HORIZONTAL,
    '|': Track.VERTICAL,
    '/': Track.SLASH,
    '\\': Track.BACKSLASH,
    '+': Track.PLUS,
    '>': Cart.RIGHT,
    '<': Cart.LEFT,
    '^': Cart.UP,
    'v': Cart.DOWN
}

SLASH_TURNS = {
    Track.SLASH: {
        Cart.RIGHT: Cart.UP,
        Cart.LEFT: Cart.DOWN,
        Cart.UP: Cart.RIGHT,
        Cart.DOWN: Cart.LEFT
    },
    Track.BACKSLASH: {
        Cart.RIGHT: Cart.DOWN,
        Cart.LEFT: Cart.UP,
        Cart.UP: Cart.LEFT,
        Cart.DOWN: Cart.RIGHT
    }
}

track = []
carts = []
with open('input/day13.txt') as read:
    for i, row in enumerate(read.readlines()):
        track.append([])
        for j, c in enumerate(row.rstrip()):
            type_ = CHAR_MAP.get(c, None)
            if isinstance(type_, Cart):
                match type_:
                    case Cart.LEFT | Cart.RIGHT:
                        track[-1].append(Track.HORIZONTAL)
                    case Cart.UP | Cart.DOWN:
                        track[-1].append(Track.VERTICAL)
                carts.append([[i, j], type_, 0])
            else:
                track[-1].append(type_)

first_crash = True
while len(carts) > 1:
    carts.sort()
    occupied = {tuple(c[0]): v for v, c in enumerate(carts)}
    still_in = {c for c in range(len(carts))}
    for v, (pos, type_, turn_num) in enumerate(carts):
        del occupied[tuple(pos)]
        match type_:
            case Cart.RIGHT:
                pos[1] += 1
            case Cart.LEFT:
                pos[1] -= 1
            case Cart.UP:
                pos[0] -= 1
            case Cart.DOWN:
                pos[0] += 1

        if tuple(pos) in occupied:
            if first_crash:
                print(f"first crash position: {pos[1]},{pos[0]}")
                first_crash = False
            still_in.remove(v)
            still_in.remove(occupied.pop(tuple(pos)))
        occupied[tuple(pos)] = v

        assert track[pos[0]][pos[1]] is not None
        match track[pos[0]][pos[1]]:
            case Track.SLASH | Track.BACKSLASH:
                carts[v][1] = SLASH_TURNS[track[pos[0]][pos[1]]][type_]
            case Track.PLUS:
                if turn_num % 3 == 0:
                    carts[v][1] = Cart((carts[v][1].value - 1) % len(Cart))
                elif turn_num % 3 == 2:
                    carts[v][1] = Cart((carts[v][1].value + 1) % len(Cart))
                carts[v][2] += 1

    carts = [carts[c] for c in still_in]

if len(carts) == 1:
    print(f"position of last cart: {carts[0][0][1]},{carts[0][0][0]}")
else:
    print("uuhh all the carts killed each other lol idt that's should happen")

File: y2018/rust/src/bin/day13.rs
use std::collections::HashMap;
use std::fs;

#[derive(Debug, Copy, Clone)]
enum Dir {
    Up,
    Down,
    Left,
    Right,
}

impl Dir {
    fn delta(&self) -> (i32, i32) {
        match self {
            Dir::Up => (-1, 0),
            Dir::Down => (1, 0),
            Dir::Left => (0, -1),
            Dir::Right => (0, 1),
        }
    }

    fn turn(&self, turn: char) -> Self {
        match turn {
            '/' => match self {
                Dir::Up => Dir::Right,
                Dir::Down => Dir::Left,
                Dir::Left => Dir::Down,
                Dir::Right => Dir::Up,
            },
            '\\' => match self {
                Dir::Up => Dir::Left,
                Dir::Down => Dir::Right,
                Dir::Left => Dir::Up,
                Dir::Right => Dir::Down,
            },
            _ => *self,
        }
    }

    fn turn_left(&self) -> Self {
        match self {
            Dir::Up => Dir::Left,
            Dir::Down => Dir::Right,
            Dir::Left => Dir::Down,
            Dir::Right => Dir::Up,
        }
    }

    fn turn_right(&self) -> Self {
        match self {
            Dir::Up => Dir::Right,
            Dir::Down => Dir::Left,
            Dir::Left => Dir::Up,
            Dir::Right => Dir::Down,
        }
    }
}

fn main() {
    let read = fs::read_to_string("../input/day13.txt").unwrap();
    let mut turns = Vec::new();
    let mut carts = Vec::new();
    for (y, r) in read.lines().enumerate() {
        let mut row_turns = vec![' '; r.len()];
        for (x, c) in r.chars().enumerate() {
            match c {
                '/' | '\\' | '+' => row_turns[x] = c,
                'v' => carts.push((Dir::Down, (y, x), 0)),
                '>' => carts.push((Dir::Right, (y, x), 0)),
                '<' => carts.push((Dir::Left, (y, x), 0)),
                '^' => carts.push((Dir::Up, (y, x), 0)),
                _ => (),
            }
        }
        turns.push(row_turns);
    }

    let mut first_crash = true;
    while carts.len() > 1 {
        carts.sort_by_key(|c| c.1);

        let mut occupied: HashMap<(usize, usize), usize> =
            carts.iter().enumerate().map(|(i, c)| (c.1, i)).collect();

        for i in 0..carts.len() {
            let (mut dir, (r, c), mut turn_num) = carts[i];
            let delta = dir.delta();
            let next = ((r as i32 + delta.0) as usize, (c as i32 + delta.1) as usize);
            let curr_ind = occupied.remove(&(r, c));
            if curr_ind.is_none() {
                continue;
            }
            let curr_ind = curr_ind.unwrap();

            if occupied.contains_key(&next) {
                occupied.remove(&next);
                if first_crash {
                    println!("loc of first crash: {},{}", next.1, next.0);
                    first_crash = false;
                }
                continue;
            }

            dir = dir.turn(turns[next.0][next.1]);
            if turns[next.0][next.1] == '+' {
                match turn_num % 3 {
                    0 => dir = dir.turn_left(),
                    2 => dir = dir.turn_right(),
                    _ => (),
                }
                turn_num += 1;
            }
            occupied.insert(next, curr_ind);
            carts[i] = (dir, next, turn_num);
        }

        carts = occupied.iter().map(|(_, ind)| carts[*ind]).collect();
    }

    if carts.is_empty() {
        println!("uh all the carts are dead,, what?");
    } else {
        let (_, left_pos, _) = carts[0];
        println!("loc of final cart: {},{}", left_pos.1, left_pos.0);
    }
}
